### PR TITLE
Split `ConfigConvert` into readers and writers

### LINKED ISSUE #92: Split ConfigConverter into two parts: for reading and for writing
There are some cases where end-user just want to read config and do not need to write it. In such cases we do not need to force to implement writer part.

### PR DESCRIPTION
**Pull request by ruippeixotog**
In this PR `ConfigConvert` is redefined as an union of two new traits: `ConfigReader` for reading objects from configs and `ConfigWriter` to write objects to configs.

`ConfigConvert` continues to exist and I think it still plays an important role in pureconfig. Most of the times users will be able to define both operations for a type, and for organization purposes it's good to define them together. However, with PR users will be able to define only one of the operations and still use operations such as `loadConfig` and `toConfig`, which is particularly useful for some particular models that only have a one-way conversion.

I opted to split the `BasicConverters` and `DerivedConverters` in readers and writers in order to mix into the companion objects of `ConfigReader` and `ConfigWriter` only the implicits that mattered to each one. However, it definitely doesn't have to be the case and I think we should leave the `ConfigConvert` definitions in pureconfig modules as they are. A `ConfigConvert` can be seen both as a `ConfigReader as a `ConfigWriter`, and in the modules we don't have any advantage in splitting them.

This change, when seen together with #171, is source-compatible with the latest version 0.6.0. If we do a release without this PR it may be possible for users to start using `BasicConverters` and `DerivedConverters` directly and this PR causes them to disappear.

Closes #92.

### CODE REVIEW

**Comment by ruippeixotog** [comment]
This PR is a WIP as it is still missing tests and documentation. I'd like to get some feedback from you before starting writing documentation. I'll be writing tests meanwhile.

**Review by melrief — DISMISSED** [review]
I'm not sure why  this is still WIP, I've read the PR and I think it's nice. I have no comments.

**Comment by ruippeixotog** [comment]
This PR is a WIP as it is still missing tests and documentation. However, as the tests for `ConfigConvert` still check both parts of the conversion, I don't think anything more is necessary for now. If you agree with the PR as it is now, I'll update the README and then we can merge this.

**Comment by melrief** [comment]
@ruippeixotog I'm concerned about possible conflicts with #178 , do you think we should wait for that to be merged before we change the documentation?

**Comment by ruippeixotog** [comment]
Good question... I guess it's better if this is merged without the documentation then and I'll do another PR later, after #178. That way no one will have much trouble dealing with merge conflicts.

**Comment by melrief** [comment]
Agreed, I'll wait for another approve and then merge

File: core/src/main/scala/pureconfig/ConvertHelpers.scala
Comment (melrief) [inline core/src/main/scala/pureconfig/ConvertHelpers.scala]: Do you think we should move reader/writer helpers in the `Reader`/`Writer` companion object? I think it could be nice to have `ConfigReader.fromString` instead of `ConverterHelper.fromStringReader`. You don't need the `Reader` at the end of `fromStringReader` if the object in which the method is defined is a `Reader` (I hope I explained myself). What do you think?

Comment (ruippeixotog) [reply core/src/main/scala/pureconfig/ConvertHelpers.scala]: Yes, that's a good idea! I'll move all the methods, but I'll leave forwarder deprecated methods in `ConfigConvert` in order not to break compatibility here, ok?

Comment (melrief) [reply core/src/main/scala/pureconfig/ConvertHelpers.scala]: Yes

**Comment by ruippeixotog** [comment]
@melrief besides the renaming you suggested, I ended up doing a greater number of changes in order to keep the code clean and with consistent names in 3a9823b.

`ConvertHelpers` doesn't have factory methods anymore, not even `ConfigConvert` ones. It ended up with nothing more than... helpers. All factory methods were moved to the adequate companion object. I prefer to have all constructors directly in the companion object and not in a separate trait, and I also like that the companion objects are not polluted anymore with methods not relating to them directly.

I tried to match the factory method name to the type of entity to build. `ConfigReader` has the `fromString` factory method family, with all try/option and non-empty variants. `ConfigWriter` has `forPrimitive`, `toDefaultString` and `toString` factory methods. `ConfigConvert` has a `viaString` factory method family like `ConfigReader`.

I left a large number of deprecated constructors in `ConfigConvert` in order not to break existing code. I hope I didn't forget any method, but everything should be there.

Let me know what you think of it now!

**Comment by melrief** [comment]
@ruippeixotog thanks for the changes, the new API is much nicer in my opinion! I don't have time to do the review but tomorrow I'll have a look in deep.

**Review by melrief — APPROVED** [review]
Nice, I like the new API for `ConfigReader` and `ConfigWriter` even more. I'm waiting for another review.

File: core/src/main/scala/pureconfig/ConfigConvert.scala
Comment (leifwickland) [inline core/src/main/scala/pureconfig/ConfigConvert.scala]: Suggestion: Add a class-level comment like:
```
/**
 * This object provides methods to create [ConfigConvert] instances.
 */
```

File: core/src/main/scala/pureconfig/ConfigReader.scala
Comment (leifwickland) [inline core/src/main/scala/pureconfig/ConfigReader.scala]: Suggestion: Add a class-level comment like:
```
/**
 * This object provides methods to create [ConfigReader] instances.
 */
```

File: core/src/main/scala/pureconfig/ConfigWriter.scala
Comment (leifwickland) [inline core/src/main/scala/pureconfig/ConfigWriter.scala]: Suggestion: Add a class-level comment like:
```
/**
 * This object provides methods to create [ConfigWriter] instances.
 */
```

File: core/src/main/scala/pureconfig/DerivedWriters.scala
Comment (leifwickland) [inline core/src/main/scala/pureconfig/DerivedWriters.scala:33]: The immortal `TODO` lives on.

File: core/src/main/scala/pureconfig/DerivedWriters.scala
Comment (leifwickland) [inline core/src/main/scala/pureconfig/DerivedWriters.scala:63]: Is this the next great zombie comment?

**Review by leifwickland — APPROVED** [review]
Kudos, @ruippeixotog! This was a huge change that seems to be have executed extremely well.
I offered a couple suggestions, but this LGTM. 👍

File: core/src/main/scala/pureconfig/DerivedWriters.scala
Comment (melrief) [reply core/src/main/scala/pureconfig/DerivedWriters.scala:33]: I think we should address old TODOs in another PR because this one is dealing with enough stuff :-)

**Review by derekmorr — APPROVED** [review]

File: core/src/main/scala/pureconfig/ConfigConvert.scala
Comment (ruippeixotog) [reply core/src/main/scala/pureconfig/ConfigConvert.scala]: Added in 5845cac.

File: core/src/main/scala/pureconfig/ConfigReader.scala
Comment (ruippeixotog) [reply core/src/main/scala/pureconfig/ConfigReader.scala]: Added in 5845cac.

File: core/src/main/scala/pureconfig/ConfigWriter.scala
Comment (ruippeixotog) [reply core/src/main/scala/pureconfig/ConfigWriter.scala]: Added in 5845cac.

File: core/src/main/scala/pureconfig/DerivedWriters.scala
Comment (ruippeixotog) [reply core/src/main/scala/pureconfig/DerivedWriters.scala:33]: I agree with @leifwickland. The comment is still relevant and unrelated to this PR.

File: core/src/main/scala/pureconfig/DerivedWriters.scala
Comment (ruippeixotog) [reply core/src/main/scala/pureconfig/DerivedWriters.scala:63]: I believe the comment is still relevant - it mentions a question for which there is still no clear answer. Though it may not be the best place to make TODOs and discussion topics visible, I don't mind having comments like this around for as long as it takes.

**Comment by ruippeixotog** [comment]
I have just fixed the merge conflicts. Unless you want to wait for @jcazevedo to review this, this is ready to merge, @melrief.

@leifwickland I believe I replied to all your comments. Please let me know if you see anything else :)

**Review by jcazevedo — APPROVED** [review]
This looks good to merge as is! Sorry for not reviewing it earlier.

**Pull request merged**
